// xo-form/dist/src/xo/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Control_1 = __importDefault(require("./Control"));
const Form_1 = __importDefault(require("./Form"));
const Group_1 = __importDefault(require("./Group"));
const Page_1 = __importDefault(require("./Page"));
const Repeat_1 = __importDefault(require("./Repeat"));
const Navigation_1 = __importDefault(require("./Navigation"));
const Context_1 = __importDefault(require("./Context"));
const Util_1 = __importDefault(require("./Util"));
const DataBinding_1 = __importDefault(require("./DataBinding"));
const PropertyMapper_1 = __importDefault(require("./PropertyMapper"));
const Validation_1 = __importDefault(require("./Validation"));
/**
 * XO - Namespace
 */
class xo {
    /**
     * XO Context
     */
    static get Context() {
        return Context_1.default;
    }
    /**
     * XO Control (```<xo-control/>```) - both Base Control for XO, and wrapping Control for other HTML elements
     */
    static get Control() {
        return Control_1.default;
    }
    /**
     * XO DataBinding - Manages dual-databinding within the form
     */
    static get DataBinding() {
        return DataBinding_1.default;
    }
    /**
     * XO Form Control (```<xo-form/>```)
     */
    static get Form() {
        return Form_1.default;
    }
    /**
     * XO Group Control (```<xo-group/>```)
     */
    static get Group() {
        return Group_1.default;
    }
    /**
     * XO Property Mapper
     */
    static get PropertyMapper() {
        return PropertyMapper_1.default;
    }
    /**
     * XO Page Control (```<xo-page/>```)
     */
    static get Page() {
        return Page_1.default;
    }
    /**
     * XO Repeat Control (```<xo-repeat/>```) - Repeats underlying structure for all items in *.items* Array.
     */
    static get Repeat() {
        return Repeat_1.default;
    }
    /**
     * Util Class - contains static helper methods
     */
    static get Util() {
        return Util_1.default;
    }
    /**
     * Navigation - Manages multi-step form navigation
     */
    static get Navigation() {
        return Navigation_1.default;
    }
    /**
     * Validation - manages form validation
     */
    static get Validation() {
        return Validation_1.default;
    }
}
window.xo = xo;
exports.default = xo;
