// xo-form/dist/src/xo/Page.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Group_1 = __importDefault(require("./Group"));
const lit_1 = require("lit");
/**
 * XO Page Control (```<xo-page/>```)
 */
class Page extends Group_1.default {
    constructor() {
        super(...arguments);
        this.hidden = true;
    }
    connectedCallback() {
        super.connectedCallback();
        this.closest("xo-form").addEventListener("page", (e) => {
            this.hidden = e.target.page !== this.index;
        });
        this.hidden = this.index !== 1;
    }
    render() {
        if (this.hidden)
            return (0, lit_1.html) ``;
        return (0, lit_1.html) `<fieldset
      ${this.hidden ? "hidden" : ""}
      data-page="${this.index}"
      class="xo-cn ${this.getContainerClasses()}"
    >
      <legend>${this.label}</legend>
      <slot></slot>
    </fieldset>`;
    }
    static get properties() {
        return {
            index: { type: Number, attribute: true },
        };
    }
}
window.customElements.define("xo-page", Page);
exports.default = Page;
